File: app/src/main/java/com/migue/zeus/expensesnotes/data/models/ExpenseCategory.java
package com.migue.zeus.expensesnotes.data.models;

import android.arch.persistence.room.Entity;
import android.arch.persistence.room.PrimaryKey;
import android.content.Context;

import com.google.gson.reflect.TypeToken;
import com.migue.zeus.expensesnotes.R;

import java.lang.reflect.Type;
import java.util.Map;

@Entity(tableName = "ExpensesCategories")
public class ExpenseCategory {
    @PrimaryKey(autoGenerate = true)
    private int Id;
    private String Name;
    private int IconId;

    public ExpenseCategory() {
    }

    public ExpenseCategory(String name, int iconId) {
        Name = name;
        this.IconId = iconId;
    }
    public ExpenseCategory(String name) {
        Name = name;
    }
    public int getId() {
        return Id;
    }

    public void setId(int id) {
        Id = id;
    }

    public String getName() {
        return Name;
    }

    public void setName(String name) {
        Name = name;
    }


    public int getIconId() {
        return IconId;
    }

    public void setIconId(int iconId) {
        IconId = iconId;
    }

    public static ExpenseCategory[] populateData(final Context context) {
        String[] rawCategories = context.getResources().getStringArray(R.array.expenses_categories);
        ExpenseCategory[] categories = new ExpenseCategory[rawCategories.length];
        int i = 0;
        //Type type = new TypeToken<Map<String, String>>(){}.getType();
        //Map<String, String> myMap = gson.fromJson("{'k1':'apple','k2':'orange'}", type);
        for (String rawCategory : rawCategories){
            categories[i] = new ExpenseCategory(rawCategory);
            i++;
        }
        return categories;
    }
}

File: app/src/main/java/com/migue/zeus/expensesnotes/data/models/ExpenseCategory.kt
package com.migue.zeus.expensesnotes.data.models

import android.arch.persistence.room.*
import android.content.Context
import android.os.Parcel
import android.os.Parcelable

import com.migue.zeus.expensesnotes.R

@Entity(
        tableName = "ExpensesCategories",
        foreignKeys = [ForeignKey(entity = Icon::class, parentColumns = ["Id"], childColumns = ["IconId"])],
        indices = [Index("IconId")])
class ExpenseCategory : BaseEntity {
    override fun getReadableName(): String? {
        return name
    }
    @ColumnInfo(name = "Name")
    var name: String? = null
    @ColumnInfo(name = "IconId")
    var iconId: Long? = null

    constructor(parcel: Parcel) : this() {
        name = parcel.readString()
        iconId = parcel.readValue(Long::class.java.classLoader) as? Long
    }

    constructor()

    @Ignore
    constructor(name: String, iconId: Long) {
        this.name = name
        this.iconId = iconId
    }
    @Ignore
    constructor(name: String) {
        this.name = name
    }


    override fun writeToParcel(parcel: Parcel, flags: Int) {
        super.writeToParcel(parcel, flags)
        parcel.writeString(name)
        parcel.writeValue(iconId)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<ExpenseCategory> {
        override fun createFromParcel(parcel: Parcel): ExpenseCategory {
            return ExpenseCategory(parcel)
        }

        override fun newArray(size: Int): Array<ExpenseCategory?> {
            return arrayOfNulls(size)
        }

        fun populateData(context: Context): Array<ExpenseCategory?> {
            val rawCategories = context.resources.getStringArray(R.array.expenses_categories)
            val categories = arrayOfNulls<ExpenseCategory>(rawCategories.size)
            var i = 0
            //Type type = new TypeToken<Map<String, String>>(){}.getType();
            //Map<String, String> myMap = gson.fromJson("{'k1':'apple','k2':'orange'}", type);
            for (rawCategory in rawCategories) {
                categories[i] = ExpenseCategory(rawCategory)
                i++
            }
            return categories
        }
    }
}

